cpu/gd32v/periph/gpio.c

```diff
@@ -184,4 +205,98 @@ void gpio_write(gpio_t pin, int value)
     }
 }
 
+#ifdef MODULE_PERIPH_GPIO_IRQ
+
+/* Forward declaration of ISR */
+static void _gpio_isr(unsigned irqn);
+
+int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
+                  gpio_cb_t cb, void *arg)
+{
+    assert(cb != NULL);
+
+    int pin_num = _pin_num(pin);
+    int port_num = _port_num(pin);
+
+    /* disable interrupts on the channel we want to edit (just in case) */
+    EXTI->INTEN &= ~(1 << pin_num);
+
+    /* configure pin as input */
+    gpio_init(pin, mode);
+
+    /* GD32V has 16 EXTI lines for GPIO interrupts, where all ports share
+     * the same EXTI line for the same pin. That means the pin PA<n> shares
+     * the EXTI line with PB<n>, PC<n>, PD<n> and PE<n>. */
+    if (exti_ctx[pin_num].cb != 0) {
+        LOG_ERROR("EXTI line for GPIO_PIN(%u, %u) is already used.\n",
+                  port_num, pin_num);
+    }
+
+    /* set callback */
+    exti_ctx[pin_num].cb = cb;
+    exti_ctx[pin_num].arg = arg;
+
+    /* enable alternate function clock for the GPIO module */
+    periph_clk_en(APB2, RCU_APB2EN_AFEN_Msk);
+
+    /* configure the EXTI channel */
+    volatile uint32_t *afio_exti_ss = &AFIO->EXTISS0 + (pin_num >> 2);
+
+    *afio_exti_ss &= ~(0xfUL << ((pin_num & 0x03) * 4));
+    *afio_exti_ss |= (uint32_t)port_num << ((pin_num & 0x03) * 4);
+
+    /* configure the active flank */
+    EXTI->RTEN &= ~(1 << pin_num);
+    EXTI->RTEN |= ((flank & 0x1) << pin_num);
+    EXTI->FTEN &= ~(1 << pin_num);
+    EXTI->FTEN |= ((flank >> 1) << pin_num);
+
+    /* clear any pending requests */
+    EXTI->PD = (1 << pin_num);
+
+    /* enable global pin interrupt */
+    unsigned irqn = (pin_num < 5) ? EXTI0_IRQn + pin_num
+                                  : ((pin_num < 10) ? EXTI5_9_IRQn
+                                                    : EXTI10_15_IRQn);
```

Comment:
That looks artsy, but makes it harder to understand what's going on.
Let's keep it simple :wink: 

```C
static inline unsigned _irq_num(unsigned pin_num)
{
    if (pin_num < 5) {
        return EXTI0_IRQn + pin_num;
    }
    if (pin_num < 10) {
        return EXTI5_9_IRQn;
    }
    return EXTI10_15_IRQn;
}
```